--- libica-1.3.6-rc3/src/icalinux.c.stale-handles	2005-10-21 17:35:03.000000000 +0200
+++ libica-1.3.6-rc3/src/icalinux.c	2006-01-12 21:01:20.000000000 +0100
@@ -1024,7 +1024,7 @@
 		NEWMECH(CKM_AES_KEY_GEN, 0, 0, CKF_GENERATE)
 	}
 
-	if (icaOpenAdapter(0, &handle)) {
+	if (icaOpenAdapter(0, &handle) == 0) { // 0 == success
 		NEWMECH(CKM_RSA_PKCS, 128, 2048,
 			CKF_HW | CKF_ENCRYPT | CKF_DECRYPT | CKF_SIGN |
 			CKF_SIGN_RECOVER | CKF_VERIFY | CKF_VERIFY_RECOVER |
