--- openssl-0.9.7a/libica-1.3.4/src/icalinux.c	2003-07-29 21:43:39.000000000 +0200
+++ openssl-0.9.7a/libica-1.3.5/src/icalinux.c	2003-10-15 22:49:05.000000000 +0200
@@ -1234,18 +1234,30 @@
                          unsigned int        outputDataLength,
                          unsigned char      *pOutputData )
 {
-
   /* check for obvious errors in parms */
   if( pOutputData == NULL){ 
     return HDDInvalidParm;
   }
-  
+
+#ifdef _LINUX_S390_
+  // We'll reuse our local copy of hAdapterHandle
+  hAdapterHandle = open("/dev/urandom", O_RDONLY);
+  if (hAdapterHandle == -1)
+    return errno;
+#endif
+
   if( read( hAdapterHandle, pOutputData, outputDataLength ) == -1 ) {
-      
+#ifdef _LINUX_S390_
+          close(hAdapterHandle);
+#endif
           return( errno );
   }
 
-  return 0;;
+#ifdef _LINUX_S390_
+  close(hAdapterHandle);
+#endif
+
+  return 0;
 }
 
 /*----------------------------------------------------------------------
