--- libica-1.3.6-rc3/src/sha1.c.no-config-h	2005-10-21 17:35:03.000000000 +0200
+++ libica-1.3.6-rc3/src/sha1.c	2006-05-11 14:18:39.000000000 +0200
@@ -355,7 +355,6 @@
 #include <string.h>
 
 #include "platform.h"
-#include <linux/config.h>
 #ifdef __s390__
 #define _LINUX_S390_ 1
 #ifdef _LINUX_S390_
--- libica-1.3.6-rc3/src/icalinux.c.no-config-h	2006-05-11 12:59:23.000000000 +0200
+++ libica-1.3.6-rc3/src/icalinux.c	2006-05-11 14:19:03.000000000 +0200
@@ -384,7 +384,6 @@
 
 #include <stdio.h>
 #include <stdlib.h>
-#include <linux/config.h>
 
 #ifdef __s390__
 #ifndef _LINUX_S390_
