--- openssl-0.9.5a/crypto/des/fcrypt.c.glibc	Thu Nov  2 15:51:29 2000
+++ openssl-0.9.5a/crypto/des/fcrypt.c	Thu Nov  2 15:51:38 2000
@@ -61,7 +61,7 @@
 void fcrypt_body(DES_LONG *out,des_key_schedule ks,
 	DES_LONG Eswap0, DES_LONG Eswap1);
 
-#if !defined(PERL5) && !defined(__FreeBSD__) && !defined(NeXT)
+#if !defined(PERL5) && !defined(__FreeBSD__) && !defined(NeXT) && !defined(__GLIBC__)
 char *crypt(const char *buf, const char *salt)
 	{
 	return(des_crypt(buf, salt));
--- openssl-0.9.5a/test/Makefile.ssl.glibc	Mon Mar 20 02:27:35 2000
+++ openssl-0.9.5a/test/Makefile.ssl	Thu Nov  2 15:22:19 2000
@@ -27,7 +27,7 @@
 
 DLIBCRYPTO= ../libcrypto.a
 DLIBSSL= ../libssl.a
-LIBCRYPTO= -L.. -lcrypto
+LIBCRYPTO= -L.. -lcrypto -lcrypt
 LIBSSL= -L.. -lssl
 
 BNTEST=		bntest
