--- openssl-engine-0.9.6b/crypto/engine/hw_ubsec.c	Fri Sep  7 19:25:04 2001
+++ openssl-engine-0.9.6b/crypto/engine/hw_ubsec.c	Fri Sep  7 19:25:25 2001
@@ -213,7 +213,9 @@
 #ifndef OPENSSL_NO_DH
 			!ENGINE_set_DH(ret, &ubsec_dh) ||
 #endif
+#ifdef HAVE_UBSEC_RAND
 	                !ENGINE_set_RAND(ret, &ubsec_rand) ||
+#endif
 			!ENGINE_set_BN_mod_exp(ret, ubsec_mod_exp) ||
 			!ENGINE_set_BN_mod_exp_crt(ret, ubsec_mod_exp_crt) ||
 			!ENGINE_set_init_function(ret, ubsec_init) ||
