These regions of memory can overlap, and often do.

diff -uNr openssl-0.9.6/ssl/s3_enc.c openssl-0.9.6/ssl/s3_enc.c
--- openssl-0.9.6/ssl/s3_enc.c	Fri Jun  1 14:40:27 2001
+++ openssl-0.9.6/ssl/s3_enc.c	Fri Jun  1 15:22:10 2001
@@ -356,7 +356,7 @@
 	if ((s->session == NULL) || (ds == NULL) ||
 		(enc == NULL))
 		{
-		memcpy(rec->data,rec->input,rec->length);
+		memmove(rec->data,rec->input,rec->length);
 		rec->input=rec->data;
 		}
 	else
diff -uNr openssl-0.9.6/ssl/t1_enc.c openssl-0.9.6/ssl/t1_enc.c
--- openssl-0.9.6/ssl/t1_enc.c	Fri Jun  1 14:40:27 2001
+++ openssl-0.9.6/ssl/t1_enc.c	Fri Jun  1 15:17:33 2001
@@ -420,7 +420,7 @@
 	if ((s->session == NULL) || (ds == NULL) ||
 		(enc == NULL))
 		{
-		memcpy(rec->data,rec->input,rec->length);
+		memmove(rec->data,rec->input,rec->length);
 		rec->input=rec->data;
 		}
 	else
