--- openssl-fips-0.9.8e/util/pl/VC-32.pl.org	2011-01-21 04:24:24.000000000 +0100
+++ openssl-fips-0.9.8e/util/pl/VC-32.pl	2011-01-21 14:03:08.531939200 +0100
@@ -150,6 +150,7 @@
 	{
 	$ex_libs.=' gdi32.lib advapi32.lib user32.lib';
 	$ex_libs.=' bufferoverflowu.lib' if ($FLAVOR =~ /WIN64/);
+	$ex_libs.=' $(EXTRA_LIBS)'
 	}
 
 # As native NT API is pure UNICODE, our WIN-NT build defaults to UNICODE,
@@ -342,6 +343,7 @@
 			$ex.=' unicows.lib' if ($FLAVOR =~ /NT/);
 			$ex.=' wsock32.lib gdi32.lib advapi32.lib user32.lib';
 			$ex.=' bufferoverflowu.lib' if ($FLAVOR =~ /WIN64/);
+			$ex.=' $(EXTRA_LIBS)'
 			}
 		$ex.=" $zlib_lib" if $zlib_opt == 1 && $target =~ /O_CRYPTO/;
 
