--- openssl-engine-0.9.6b/crypto/engine/hw_ubsec.c	Tue Aug 28 16:58:55 2001
+++ openssl-engine-0.9.6b/crypto/engine/hw_ubsec.c	Tue Aug 28 17:01:26 2001
@@ -329,7 +329,7 @@
 	t_UBSEC_math_accelerate_ioctl *p11;
 	t_UBSEC_rng_ioctl *p12;
 	t_UBSEC_max_key_len_ioctl *p13;
-	int fd = 0;
+	int fd;
 
 	if(ubsec_dso != NULL)
 		{
@@ -398,6 +398,7 @@
 	if (p_UBSEC_max_key_len_ioctl(fd, &max_key_len) != 0)
 	        {
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
+		p_UBSEC_ubsec_close(fd);
 		goto err;
 		}
 	
@@ -495,8 +496,10 @@
 
 	if (r_len > max_key_len) 
 	        {
+		ENGINE *e;
 		ENGINEerr(ENGINE_F_UBSEC_MOD_EXP, ENGINE_R_SIZE_TOO_LARGE_OR_TOO_SMALL);
-		ret = BN_mod_exp(r, a, p, m, ctx);
+	        e = ENGINE_openssl();
+	        ret = e->bn_mod_exp(r, a, p, m, ctx);
 		goto err;
 		} 
 
@@ -509,8 +512,10 @@
 
 	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0) 
 	        {
-		fd = 0;
+		ENGINE *e;
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
+	        e = ENGINE_openssl();
+	        ret = e->bn_mod_exp(r, a, p, m, ctx);
 		goto err;
 		}
 
@@ -520,12 +525,14 @@
 				      (unsigned char *)p->d, BN_num_bits(p), 
 				      (unsigned char *)r->d, &r_len) != 0)
 	        {
+		/* Hardware's a no go, failover to software */
+		ENGINE *e;
+
 		ENGINEerr(ENGINE_F_UBSEC_MOD_EXP, ENGINE_R_REQUEST_FAILED);
 		p_UBSEC_ubsec_close(fd);
 
-		/* Hardware's a no go, failover to software */
-
-		ret = BN_mod_exp(r, a, p, m, ctx);
+	        e = ENGINE_openssl();
+	        ret = e->bn_mod_exp(r, a, p, m, ctx);
 		goto err;
 		}
 	
@@ -555,14 +562,11 @@
 
 	if (y_len > max_key_len) 
 	        {
-		ENGINEerr(ENGINE_F_UBSEC_RSA_MOD_EXP_CRT, ENGINE_R_SIZE_TOO_LARGE_OR_TOO_SMALL);
-
-		{ 
 	        ENGINE *e;
-	    
+
+		ENGINEerr(ENGINE_F_UBSEC_RSA_MOD_EXP_CRT, ENGINE_R_SIZE_TOO_LARGE_OR_TOO_SMALL);
 	        e = ENGINE_openssl();
 	        ret = e->bn_mod_exp_crt(r, a, p, q, dp, dq, qinv, ctx);
-	        }
 
 		goto err;
 	        }
@@ -575,10 +579,12 @@
 
 	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0) 
 	        {
-		fd = 0;
+	        ENGINE *e;
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
+	        e = ENGINE_openssl();
+	        ret = e->bn_mod_exp_crt(r, a, p, q, dp, dq, qinv, ctx);
 		goto err;
-	        }
+		}
 
 	if (p_UBSEC_rsa_mod_exp_crt_ioctl(fd,
 		(unsigned char *)a->d, BN_num_bits(a), 
@@ -589,18 +595,12 @@
 		(unsigned char *)q->d, BN_num_bits(q),
 		(unsigned char *)r->d,  &y_len) != 0) 
 	        {
+	        ENGINE *e;
 		p_UBSEC_ubsec_close(fd);
 		ENGINEerr(ENGINE_F_UBSEC_RSA_MOD_EXP_CRT, ENGINE_R_UNIT_FAILURE);
-	  
-	        /* Hardware's a no go, failover to software */
-
-	        { 
-	        ENGINE *e;
-	    
 	        e = ENGINE_openssl();
 	        ret = e->bn_mod_exp_crt(r, a, p, q, dp, dq, qinv, ctx);
-	        }
-	  
+
 	        goto err;
 	        }
 
@@ -632,10 +632,9 @@
 	if ((BN_num_bits(rsa->p) > (max_key_len / 2)) && 
 	    (BN_num_bits(rsa->q) > (max_key_len / 2 )))
 	        {
-		const RSA_METHOD *m = NULL;
-
-		m = RSA_PKCS1_SSLeay();
-		ret = (*m->rsa_mod_exp)(r0, I, rsa);
+		const RSA_METHOD *meth;
+		meth = RSA_PKCS1_SSLeay();
+		ret = meth->rsa_mod_exp(r0, I, rsa);
 	        } 
 	else 
 	        {
@@ -700,8 +699,9 @@
 
 	if (BN_num_bits(m) > max_key_len) 
 	        {
-		const RSA_METHOD *meth = RSA_PKCS1_SSLeay();
-		ret = (*meth->bn_mod_exp)(r, a, p, m, ctx, m_ctx);
+		const RSA_METHOD *meth;
+		meth = RSA_PKCS1_SSLeay();
+		ret = meth->bn_mod_exp(r, a, p, m, ctx, m_ctx);
 		} 
 	else 
 #endif
@@ -742,11 +742,14 @@
 		goto err;
 	} 
 
-	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0) {
-		fd = 0;
+	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0)
+		{
+		const DSA_METHOD *meth;
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
-		return 0;
-	}
+		meth = DSA_OpenSSL();
+		ret = meth->dsa_do_sign(dgst, dlen, dsa);
+		goto err;
+		}
 
 	if (p_UBSEC_dsa_sign_ioctl(fd, 
 				   0, /* compute hash before signing */
@@ -759,15 +762,14 @@
 				   (unsigned char *)r->d, &r_len,
 				   (unsigned char *)s->d, &s_len ) != 0) 
 	        {
+		/* Hardware's a no go, failover to software */
+		const DSA_METHOD *meth;
+
 		ENGINEerr(ENGINE_F_UBSEC_DSA_SIGN, ENGINE_R_REQUEST_FAILED);
 		p_UBSEC_ubsec_close(fd);
 
-		/* Hardware's a no go, failover to software */
-
-		{
-		const DSA_METHOD *meth = DSA_OpenSSL();
+		meth = DSA_OpenSSL();
 		ret = meth->dsa_do_sign(dgst, dlen, dsa);
-		}
 
 		goto err;
 	}
@@ -817,11 +819,13 @@
 	d_len = p_UBSEC_ubsec_bytes_to_bits((unsigned char *)dgst, dgst_len);
 
 	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0) 
-        {
-		fd = 0;
+        	{
+		const DSA_METHOD *meth;
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
+		meth = DSA_OpenSSL();
+		ret = meth->dsa_do_verify(dgst, dgst_len, sig, dsa);
 		goto err;
-	}
+		}
 
 	if (p_UBSEC_dsa_verify_ioctl(fd, 
 				     0, /* compute hash before signing */
@@ -834,15 +838,14 @@
 				     (unsigned char *)sig->s->d, BN_num_bits(sig->s),
 				     (unsigned char *)v.d, &v_len) != 0) 
 	        {
+		/* Hardware's a no go, failover to software */
+		const DSA_METHOD *meth;
+		
 		ENGINEerr(ENGINE_F_UBSEC_DSA_VERIFY , ENGINE_R_REQUEST_FAILED);
 		p_UBSEC_ubsec_close(fd);
 
-		/* Hardware's a no go, failover to software */
-
-		{
-		const DSA_METHOD *meth = DSA_OpenSSL();
+		meth = DSA_OpenSSL();
 		ret = meth->dsa_do_verify(dgst, dgst_len, sig, dsa);
-		}
 
 		goto err;
 		}
@@ -882,8 +885,10 @@
 	  
 	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0) 
 	        {
-		fd = 0;
+		const DH_METHOD *meth;
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
+		meth = DH_OpenSSL();
+		ret = meth->compute_key(key, pub_key, dh);
 		goto err;
 		}
 	  
@@ -893,16 +898,13 @@
 					       (unsigned char *)dh->p->d, BN_num_bits(dh->p),
 					       key, &k_len) != 0)
 	        {
+	        /* Hardware's a no go, failover to software */
+		const DH_METHOD *meth;
 		ENGINEerr(ENGINE_F_UBSEC_DH_COMPUTE_KEY, ENGINE_R_REQUEST_FAILED);
 		p_UBSEC_ubsec_close(fd);
 		
-	        /* Hardware's a no go, failover to software */
-
-		{
-		const DH_METHOD *meth;
 		meth = DH_OpenSSL();
 		ret = meth->compute_key(key, pub_key, dh);
-		}
 		
 		goto err;
 		}
@@ -959,8 +961,10 @@
 
 	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0) 
 	        {
-	        fd = 0;
+		const DH_METHOD *meth;
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
+      		meth = DH_OpenSSL();
+		ret = meth->generate_key(dh);
 		goto err;
 		}
   
@@ -971,16 +975,14 @@
 						  (unsigned char *)dh->p->d, BN_num_bits(dh->p),
 						  0, 0, random_bits) != 0) 
 	        {
+		/* Hardware's a no go, failover to software */
+		const DH_METHOD *meth;
+
 		ENGINEerr(ENGINE_F_UBSEC_DH_COMPUTE_KEY, ENGINE_R_REQUEST_FAILED);
 		p_UBSEC_ubsec_close(fd);
 
-		/* Hardware's a no go, failover to software */
-
-		{
-		const DH_METHOD *meth;
       		meth = DH_OpenSSL();
 		ret = meth->generate_key(dh);
-		}
 
 		goto err;
 		}
@@ -1006,7 +1008,12 @@
 
 	if ((fd = p_UBSEC_ubsec_open(UBSEC_KEY_DEVICE_NAME)) <= 0) 
 	        {
+		const RAND_METHOD *meth;
 		ENGINEerr(ENGINE_F_UBSEC_INIT, ENGINE_R_UNIT_FAILURE);
+		num = p_UBSEC_ubsec_bits_to_bytes(num);
+		meth = RAND_SSLeay();
+		meth->seed(buf, num);
+		ret = meth->bytes(buf, num);
 	        goto err;
 		}
 
@@ -1017,19 +1024,16 @@
 			      buf,
 			      &num) != 0)
 	        {
+		/* Hardware's a no go, failover to software */
+		const RAND_METHOD *meth;
+
 		ENGINEerr(ENGINE_F_UBSEC_RNG_BYTES, ENGINE_R_REQUEST_FAILED);
 		p_UBSEC_ubsec_close(fd);
 
-		/* Hardware's a no go, failover to software */
-
 		num = p_UBSEC_ubsec_bits_to_bytes(num);
-
-		{
-		const RAND_METHOD *meth;
 		meth = RAND_SSLeay();
 		meth->seed(buf, num);
 		ret = meth->bytes(buf, num);
-		}
 
 	        goto err;
 		}
