--- openssl-0.9.5a/crypto/bn/bn.h	Wed Jun 28 07:27:06 2000
+++ openssl-0.9.5a/crypto/bn/bn.h	Wed Jun 28 07:27:19 2000
@@ -72,11 +72,13 @@
 #undef BN_LLONG /* experimental, so far... */
 #endif
 
+#if 0
 #define BN_MUL_COMBA
 #define BN_SQR_COMBA
 #define BN_RECURSION
 #define RECP_MUL_MOD
 #define MONT_MUL_MOD
+#endif
 
 /* This next option uses the C libraries (2 word)/(1 word) function.
  * If it is not defined, I use my C version (which is slower).
--- openssl-0.9.5a/include/openssl/opensslconf.h	Wed Jun 28 07:26:51 2000
+++ openssl-0.9.5a/include/openssl/opensslconf.h	Thu Jun 15 23:02:41 2000
@@ -74,9 +74,9 @@
 /* The prime number generation stuff may not work when
  * EIGHT_BIT but I don't care since I've only used this mode
  * for debuging the bignum libraries */
+#define SIXTY_FOUR_BIT_LONG
-#undef SIXTY_FOUR_BIT_LONG
 #undef SIXTY_FOUR_BIT
+#undef THIRTY_TWO_BIT
-#define THIRTY_TWO_BIT
 #undef SIXTEEN_BIT
 #undef EIGHT_BIT
 #endif
